### PR TITLE
All Gather Once (FSDP Zero-one)

### PR DESCRIPTION
**Pull request by wei879-100**
# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/configs/base.yml [hunk 1]
```diff
@@ -186,6 +186,9 @@ pipeline_delay_activation_forwarding: False # This delays the activation forward
 # the communication and compute in each iteration are now independent. However this comes at the cost of doubling the pipeline bubble,
 # and you must set the number of microbatches to at least 2 * num_stages (the minimum 2 * num_stages is set by default with this delay).
 
+model_fsdp_ag_once: False # This controls whether the Zero-1 optimization is active.
```
Comment (gobbleturk) [inline MaxText/configs/base.yml:189]: how does this interplay with pipeline_fsdp_ag_once? If you set the model one to true, what happens when we set pipeline_fsdp_ag_once to true/false?

File: MaxText/layers/models.py [hunk 2]
```diff
@@ -804,4 +804,175 @@ def __call__(
...
\ No newline at end of file
+
+class ZeroOneTransformer(nn.Module):
+  """
+  A wrapper for the base Transformer model designed to implement the Zero-1
+  FSDP optimization.
+
+  The goal of this optimization is to reduce communication overhead. In the standard
+  FSDP implementation, an all-gather operation on the model weights is performed twice
+  for each microbatch (once for the forward pass, once for the backward pass).
```
Comment (gobbleturk) [inline MaxText/layers/models.py]: microbatch -> gradient accumulation microbatch

File: MaxText/configs/base.yml [hunk 3]
```diff
@@ -186,6 +186,9 @@ pipeline_delay_activation_forwarding: False # This delays the activation forward
...
+# This is a memory/time tradeoff - True: Use ZeroOneTransformer to gather weights once per gradient step.
```
Comment (gobbleturk) [inline MaxText/configs/base.yml]: Add - This is zero one sharding

File: MaxText/configs/base.yml [hunk 4]
```diff
@@ -186,6 +186,9 @@ pipeline_delay_activation_forwarding: False # This delays the activation forward
 # the communication and compute in each iteration are now independent. However this comes at the cost of doubling the pipeline bubble,
 # and you must set the number of microbatches to at least 2 * num_stages (the minimum 2 * num_stages is set by default with this delay).
 
+model_fsdp_ag_once: False # This controls whether the Zero-1 optimization is active.
+# This is a memory/time tradeoff - True: Use ZeroOneTransformer to gather weights once per gradient step. 
+# False: Use the standard Transformer, which gathers for each microbatch's fwd/bwd pass.
```
Comment (gobbleturk) [inline MaxText/configs/base.yml]: Add - this is zero 3 sharding

File: MaxText/layers/models.py [hunk 5]
```diff
@@ -804,4 +804,175 @@ def __call__(
...
+      
+      partition_spec_tree = jax.tree.map(get_partition_spec_leaf, pytree, is_leaf=_is_leaf)
+      return partition_spec_tree
+    # Run the function to get the full tree of partition specs.
+    partition_spec_with_extra_layer = get_partition_spec(weights)
+    # The actual model parameters are nested under the 'layers' key. This extracts them.
+    partition_spec = {"params": partition_spec_with_extra_layer["params"]["layers"]}
+    return partition_spec
+  
+  def get_physical_spec_no_fsdp(self, full_logical):
```
Comment (gobbleturk) [inline MaxText/layers/models.py]: Can this implementation be shared with the PP version as well (https://github.com/AI-Hypercomputer/maxtext/blob/f52c26c17b53b36bf1e741e19550cade98772737/MaxText/layers/pipeline.py#L561)

File: MaxText/layers/models.py [hunk 6]
```diff
@@ -804,4 +804,175 @@ def __call__(
...
+        return named_sharding
+      return jax.tree.map(_remove_fsdp_from_partition_spec, sharding_tree)
+
+    # Convert the high-level logical spec to a physical one using default rules.
+    physical = nn.logical_to_mesh_sharding(full_logical, mesh=self.mesh, rules=self.config.logical_axis_rules)
+    # Apply the function to remove the FSDP sharding, defining our target layout.
+    physical_no_fsdp = remove_fsdp_sharding(physical)
+    return physical_no_fsdp
+  
+  def all_gather_over_fsdp(self,sharding_info):
```
Comment (gobbleturk) [inline MaxText/layers/models.py]: Can this function also be shared (if so can we define it in maxtext_utils)

File: MaxText/layers/models.py (hunk 5 above)
Comment (gobbleturk) [reply MaxText/layers/models.py]: If so lets move it to maxtext_utils and have both PP and this one call the shared one in maxtext_utils

**Comment by gobbleturk** [comment]
Can we also add a unit test (integration test) with this feature on in [train_tests](https://github.com/AI-Hypercomputer/maxtext/blob/f52c26c17b53b36bf1e741e19550cade98772737/MaxText/tests/integration_tests/train_tests.py#L147)

in your PR description you can link the profiles before and after (ideally with pdb=1 to make the feature more obvious)

File: MaxText/configs/base.yml (hunk 1 above)
Comment (wei879-100) [reply MaxText/configs/base.yml:189]: Xplane:
model_fsdp_ag_once: True; pipeline_fsdp_ag_once: True:  
https://xprof.corp.google.com/trace_viewer/fanwei-3085710629162492243?hosts=t1v-n-aade02bb-w-0&host_index=0&trace_filter_config={}

model_fsdp_ag_once: False; pipeline_fsdp_ag_once: True:  
https://xprof.corp.google.com/trace_viewer/fanwei-3085710629162494718?hosts=t1v-n-aade02bb-w-0&host_index=0&trace_filter_config={}&view_start=-343.119&view_end=2636.957

both as True takes longer training time. Although the time for the fwd/bwd pass is shorter, there is a little "tail" after the bwd pass, some extra copy did take longer. The bug has been created here: b/427789745

File: MaxText/configs/base.yml [hunk 7]
```diff
@@ -186,7 +186,10 @@ pipeline_delay_activation_forwarding: False # This delays the activation forward
 # the communication and compute in each iteration are now independent. However this comes at the cost of doubling the pipeline bubble,
 # and you must set the number of microbatches to at least 2 * num_stages (the minimum 2 * num_stages is set by default with this delay).
 
-pipeline_fsdp_ag_once: False # If set to true then all gather all of the weights over FSDP before the first pipeline iteration.
+model_fsdp_ag_once: False # This is ZeRO-one sharding that controls whether the Zero-one optimization is active.
+# This is a memory/time tradeoff - True: Use ZeroOneTransformer to gather weights once per gradient step. 
+# False: Use the standard Transformer, which gathers for each microbatch's fwd/bwd pass.
+pipeline_fsdp_ag_once: False # This is ZeRO-three sharding. If set to true then all gather all of the weights over FSDP before the first pipeline iteration.
```
Comment (gobbleturk) [inline MaxText/configs/base.yml]: I wouldn't change this comment

File: MaxText/layers/models.py [hunk 8]
```diff
@@ -804,4 +803,74 @@ def __call__(
         image_embeddings=image_embeddings,
     )
     return logits
-  
\ No newline at end of file
+
+
+class ZeroOneTransformer(nn.Module):  # maybe change ZeroOneWrapper()
```
Comment (gobbleturk) [inline MaxText/layers/models.py]: We can keep it as ZeroOneTransformer - I would remove the # maybe change ZeroOneWrapper comment

Comment (gobbleturk) [reply MaxText/layers/models.py]: Especially as implemented now its hardcoded to the Transformer, if this was a generic wrapper it should take the wrapped model (e.g. Transformer) as input

File: MaxText/tests/integration_tests/train_tests.py [hunk 9]
```diff
@@ -334,6 +334,38 @@ def test_gpu_cudnn_flash_jax(self):
...
+    train_main(TrainTests.CONFIGS["base"] + ["model_fsdp_ag_once=True"])
+
+  @pytest.mark.integration_test
+  @pytest.mark.gpu_only
+  def test_gpu_synthetic_model_ag_once(self):
+    train_main(TrainTests.CONFIGS["synthetic"] + ["model_fsdp_ag_once=True"])
+
+  @pytest.mark.integration_test
+  @pytest.mark.gpu_only
+  def test_gpu_context_parallel_ag_once(self):
```
Comment (gobbleturk) [inline MaxText/tests/integration_tests/train_tests.py]: I would remove this test, I don't see ithe value its adding, you are already adding a GPU test above

File: MaxText/configs/base.yml (hunk 7 above)
Comment (wei879-100) [reply MaxText/configs/base.yml]: I put the comment to the wrong line previously, should be after "False". Revised

File: MaxText/train.py [hunk 10]
```diff
@@ -602,7 +602,13 @@ def setup_mesh_and_model(config, devices=None):
...
   # Model and Optimizer definition
   quant = quantizations.configure_quantization(config)
-  model = Transformer(config, mesh, quant=quant)
+  # Determine which model class to use based on the config flag.
+  if config.model_fsdp_ag_once:
+    # Instantiate the Zero-1 wrapper model if the optimization is enabled.
+    model = ZeroOneTransformer(config, mesh, quant=quant)
+  else:
+    # Otherwise, use the standard memory-efficient Transformer model.
+    model = Transformer(config, mesh, quant=quant)
```
Comment (bvandermoon) [inline MaxText/train.py]: Could you move this logic to a separate method (and possibly a separate file - maybe `maxtext_utils` if there isn't a better spot)? Then this call could just be something like `model = get_transformer_model(...)`

Comment (wei879-100) [reply MaxText/train.py]: Thanks for your suggestion! I put the call on the train_utils.py instead. Please check again. Thanks!

**Comment by hx89** [comment]
Thanks for the PR! 

I was testing this PR on llama3 70b mode and the profile had two unexpected behavior:
1. all-gather in the fprop is in fp32 instead of bf16
2. bprop has all-reduce instead of reduce-scatter

Did you see these behavior in your test?

And could you please share the profile you got into a public repo so that I can compare? The original link is Google internal and I don't have access.
